### PR TITLE
Add flutter_build package for codegen and version sync with tool

### PR DESCRIPTION
**Pull request by jonahwilliams**
Includes a single build which can produce an .app.dill and a modified .packages file using the frontend server. The configuration options mirror those available in the flutter_tool and can be configured with defines similar to blaze.

When a user takes a dev_dependency on flutter_build, they can run the following command to invoke this builder. (This is just a building block of codegen, not the final workflow)

```
exec flutter pub pub run build_runner build \
--define "flutter_build|kernel=disabled=false" \
--define "flutter_build|kernel=aot=false" \
--define "flutter_build|kernel=linkPlatformKernelIn=false" \
--define "flutter_build|kernel=trackWidgetCreation=false" \
--define "flutter_build|kernel=targetProductVm=false" \
--define "flutter_build|kernel=mainPath=/Users/user/Documents/testbed/lib/main.dart" \
--define "flutter_build|kernel=packagesPath=/Users/user/Documents/testbed/.packages" \
--define "flutter_build|kernel=sdkRoot=/Users/user/Documents/Flutter/flutter/bin/cache/artifacts/engine/common/flutter_patched_sdk" \
--define "flutter_build|kernel=frontendServerPath=/Users/user/Documents/Flutter/flutter/bin/cache/artifacts/engine/darwin-x64/frontend_server.dart.snapshot" \
--define "flutter_build|kernel=engineDartBinaryPath=/Users/user/Documents/Flutter/flutter/bin/cache/dart-sdk/bin/dart" \
```

Note: This package should not be synced internally.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Apparently the update-packages went wrong

**Comment by jonahwilliams** [comment]
cc @gspencergoog , do you recognize the issue that docs is having here?

File: packages/flutter_build/lib/src/kernel_builder.dart [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+  /// Whether the `trackWidgetCreation` flag is provided to the frontend
+  /// server.
+  final bool trackWidgetCreation;
+
+  /// Whether to provide the Dart product define to the frontend server.
+  final bool targetProductVm;
+  final bool linkPlatformKernelIn;
+  final String incrementalCompilerByteStorePath;
+  final List<String> extraFrontEndOptions;
```
Comment (yjbanov) [inline packages/flutter_build/lib/src/kernel_builder.dart]: Why not document the last three fields? You're almost there! :smile:

File: packages/flutter_build/lib/src/kernel_builder.dart [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+  @override
+  Map<String, List<String>> get buildExtensions => const <String, List<String>>{
+    '.dart': <String>[_kFlutterDillOutputExtension, _kPackagesExtension],
+  };
+
+  @override
+  Future<void> build(BuildStep buildStep) async {
+    // Do not run builder if has been disabled or if this asset does not
```
Comment (yjbanov) [inline packages/flutter_build/lib/src/kernel_builder.dart]: nit: if _it_ has been disabled

File: packages/flutter_build/lib/src/kernel_builder.dart [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+      server.stdout
+        .transform<String>(utf8.decoder)
+        .transform<String>(const LineSplitter())
+        .listen(_stdoutHandler.handler);
+      await server.exitCode;
+      await _stdoutHandler.compilerOutput.future;
+      await buildStep.writeAsBytes(outputId, await outputFile.readAsBytes());
+      await buildStep.writeAsBytes(packagesOutputId, await packagesFile.readAsBytes());
+    } catch (err, stackTrace) {
+      log.shout('frontend server failed to start: $err, $stackTrace');
```
Comment (yjbanov) [inline packages/flutter_build/lib/src/kernel_builder.dart]: Is logging enough when this happens? What does `shout` do?

File: packages/flutter_build/lib/src/kernel_builder.dart [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+class StdoutHandler {
+  StdoutHandler() {
+    reset();
+  }
+
+  bool compilerMessageReceived = false;
+  String boundaryKey;
+  Completer<CompilerOutput> compilerOutput;
```
Comment (yjbanov) [inline packages/flutter_build/lib/src/kernel_builder.dart]: I think these fields can be private. They are not used outside this class. Or maybe the whole class should be private?

**Review by yjbanov — APPROVED** [review]
Only a bunch of nits.

File: packages/flutter_build/lib/src/kernel_builder.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_build/lib/src/kernel_builder.dart]: Shout leaves an error log in .dart_tool/build/generated/<<build-id>>/<<package>>/<<path>> that can be read out later. I believe this also causes the build to complete with an error, will double-check

File: packages/flutter_build/pubspec.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
+name: flutter_build
+
+environment:
+  # The pub client defaults to an <2.0.0 sdk constraint which we need to explicitly overwrite.
+  sdk: ">=2.0.0-dev.68.0 <3.0.0"
+
+dependencies:
+  # To update these, use "flutter update-packages --force-upgrade".
```
Comment (tvolkert) [inline packages/flutter_build/pubspec.yaml]: Should we add `flutter_build` here?

https://github.com/flutter/flutter/blob/d0595f6bbd460c0835a404307a5252758c5fb795/packages/flutter_tools/lib/src/commands/update_packages.dart#L122

**Comment by tvolkert** [comment]
Tests?

**Comment by jonahwilliams** [comment]
> Tests?

Will work on tests

**Comment by jonahwilliams** [comment]
Re:offline discussion, to make sure we have reasonable coverage I will add an integration test that triggers code generation and then runs the resulting code

File: packages/flutter_build/pubspec.yaml (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_build/pubspec.yaml]: Done

**Review by tvolkert — APPROVED** [review]

**Pull request merged**